### PR TITLE
Ob 773 migrate ottoman

### PR DESCRIPTION
**Pull request by xdolnak**

### CODE REVIEW

**Comment by xdolnak** [comment]
TODO for this iteration:

Fix linter - DONE
Fix publishing after merge to master

DISCUSS for this iteration:

Commit build folder? When to build js files? (Before push?)
Add comments for type checks?
Ts-ignores

TODO next iteration:

Finish soft refactor = require -> import, var -> let, const
Move code to classes, add proper types

File: .circleci/config.yml
Comment (Properko) [inline .circleci/config.yml]: ts-lint is deprecated, upgrade eslint, and pull in `@typescript-eslint/parser` and whatever else is needed

File: tsconfig.json
Comment (Properko) [inline tsconfig.json:1]: > can we use `"strict": true` (at least with the exception of `"noImplicitAny": false`) or will that be part of the next iteration?

Disregard this pls, probably would make more sense as part of the next iteration.

File: lib/ottoman.ts
Comment (Properko) [inline lib/ottoman.ts]: let's remove the comment instead

**Review by Properko — APPROVED** [review]
LGTM. Thanks for keeping this PR as simple as possible.
Agree with your breakdown into iterations.

Let's get the linter working here and add breaking rules to ignored, pls.
Then remove the ignores and address the problems in code in the next iteration, as well.

nit: I'd prefer `@ts-expect-error`s instead of `@ts-ignore`s. What's nice is, that it has to be removed, once irrelevant.

Personally, I'd prefer not to commit the built files. Makes a lot of noise in the changes of the PRs.
If the need arises, they can always be built on-demand even if referencing the git repo as a dependency, and as for published packages go, of course they should be included there.

File: tsconfig.json
Comment (xdolnak) [reply tsconfig.json:1]: "strict": true,
    702 errors, maybe in next iteration

**Comment by xdolnak** [comment]
> LGTM. Thanks for keeping this PR as simple as possible.
> Agree with your breakdown into iterations.
> 
> Let's get the linter working here and add breaking rules to ignored, pls.
> Then remove the ignores and address the problems in code in the next iteration, as well.
> 
> nit: I'd prefer `@ts-expect-error`s instead of `@ts-ignore`s. What's nice is, that it has to be removed, once irrelevant.
> 
> Personally, I'd prefer not to commit the built files. Makes a lot of noise in the changes of the PRs.
> If the need arises, they can always be built on-demand even if referencing the git repo as a dependency, and as for published packages go, of course they should be included there.

I applied your proposed changes except for the last comment about built files. If it is not committed it causes issues with cirlce-ci crashing on the build when branch is used in package.json. I tried to use postinstall, so circle-ci builds it for itself, but it crashes.

**Comment by Properko** [comment]
As mentioned, I'm against having built artifacts in the repo.
This is a workaround for a problem that has other solutions, e.g. you can always build on-demand when pulling package directly from GitHub. In prod you should always pull a published package anyway.

CONS to me seem to be that it causes unncessary changes in every PR and you as both the reviewer and dev have to make sure they're always commited up-to-date with the source files, leading to bad DX.
OTOH, as you're the future maintainer, this decision is ultimately up to you.

**Comment by xdolnak** [comment]
> As mentioned, I'm against having built artifacts in the repo.
> This is a workaround for a problem that has other solutions, e.g. you can always build on-demand when pulling package directly from GitHub. In prod you should always pull a published package anyway.
> 
> CONS to me seem to be that it causes unncessary changes in every PR and you as both the reviewer and dev have to make sure they're always commited up-to-date with the source files, leading to bad DX.
> OTOH, as you're the future maintainer, this decision is ultimately up to you.

We will keep it this way for the time being and change it when there will be less work to do to make sure every push commit is run with circle-ci without fails

**Comment by Properko** [comment]
> We will keep it this way for the time being and change it when there will be less work to do to make sure every push commit is run with circle-ci without fails

Agreed, after playing around for a bit with [this branch](https://github.com/etops/node-ottoman/commits/ignore-build-artifacts) pulled into ow-back as a dependency, I see there are more issues with building this repo as a package dependency on-demand.
- Chai version discrepancy causes issues with typescript types preventing build - needs to have the same chai version as ow-back (or maybe it should suffice not to serve the test files)
- Node version discrepancy causes `npm ERR! cb.apply is not a function`
- files entry in package.json causes lib folder not to be downloaded (should be fixed in this PR IMHO)

**Pull request merged**
